sycl/include/sycl/types.hpp

```diff
@@ -642,19 +564,19 @@ template <typename Type, int NumElements> class vec {
                 typename std::enable_if_t<std::is_same_v<vector_t_, vector_t> &&
                                           !std::is_same_v<vector_t_, DataT>>>
   constexpr vec(vector_t openclVector) {
-    if constexpr (!IsUsingArrayOnDevice) {
-      m_Data = openclVector;
-    } else {
-      m_Data = bit_cast<DataType>(openclVector);
-    }
+    m_Data = detail::BitCast<vector_t, DataType>(openclVector);
   }
 
+  /* @SYCL2020
+   * Available only when: compiled for the device.
+   * Converts this SYCL vec instance to the underlying backend-native vector
+   * type defined by vector_t.
+   */
   operator vector_t() const {
-    if constexpr (!IsUsingArrayOnDevice) {
+    if constexpr (NumElements == 1) {
       return m_Data;
     } else {
-      auto ptr = bit_cast<const vector_t *>((&m_Data)->data());
-      return *ptr;
+      return detail::BitCast<DataType, vector_t>(m_Data);
```

Comment:
We don't need constexpr here, so normal `sycl::bit_cast` without UB could be used. Note that `enum`-based implementation of `detail::BitCast` is still UB technically.